topdown/reachable.go

```diff
@@ -56,6 +68,71 @@ func builtinReachable(bctx BuiltinContext, args []*ast.Term, iter func(*ast.Term
 	return iter(ast.NewTerm(reached))
 }
 
+// pathBuilder is called recursively to build an array of paths that are reachable from the root
+func pathBuilder(graph ast.Object, root *ast.Term, path []*ast.Term, paths []*ast.Term, reached ast.Set) []*ast.Term {
+	if edges := graph.Get(root); edges != nil {
+		path = append(path, root)
+
+		if numberOfEdges(edges) >= 1 {
+			foreachVertex(edges, func(neighbor *ast.Term) {
+				if reached.Contains(neighbor) {
+					// If we've already reached this node, return current path (avoid infinite recursion)
+					paths = append(paths, ast.ArrayTerm(path...))
+				} else {
+					reached.Add(root)
+					paths = pathBuilder(graph, neighbor, path, paths, reached)
+				}
+			})
+		} else {
+			paths = append(paths, ast.ArrayTerm(path...))
+		}
+	}
+
+	return paths
+}
+
+func builtinReachablePaths(bctx BuiltinContext, args []*ast.Term, iter func(*ast.Term) error) error {
+	// Return the empty set if the first argument is not an object.
+	graph, ok := args[0].Value.(ast.Object)
```

Comment:
The existing `builtinReachable` function was returning an empty set, too, so I assumed this would want to do the same. I'll modify this to return an error, as per your suggestion.

Reply:
@srenatus - I'm unsure of which helper you're referring to, though: it looks like the other built-ins return an error with `fmt.Errorf`; could you point me in the right direction?

Reply:
I wonder if this is on the right track...
```
		return builtins.NewOperandTypeErr(1, args[0].Value, ast.TypeName(types.Object{}))
```

Reply:
Sorry I was on mobile 😬  `builtins.ArrayOperand` is what I meant: https://github.com/open-policy-agent/opa/blob/27c8d75d3528dc8a04e716d5455f97ee7fb537bc/topdown/builtins/builtins.go#L152-L157. If it returns an error, passing that along from the built-in function is fine.

Reply:
🤦 Not ArrayOperand. ObjectOperand. https://github.com/open-policy-agent/opa/blob/27c8d75d3528dc8a04e716d5455f97ee7fb537bc/topdown/builtins/builtins.go#L142